src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+
+                var rewrittenArguments = VisitArguments(
+                    appendCall.Syntax,
+                    appendCall.Arguments,
+                    appendCall.Method,
+                    appendCall.ArgsToParamsOpt,
+                    argRefKindsOpt,
+                    ref appendReceiver,
+                    out ArrayBuilder<LocalSymbol>? temps);
+
+                var rewrittenAppendCall = MakeArgumentsAndCall(
                     appendCall.Syntax,
-                    builderTemp,
+                    appendReceiver,
                     appendCall.Method,
-                    rewrittenArgs,
+                    rewrittenArguments,
                     appendCall.ArgumentRefKindsOpt,
                     appendCall.Expanded,
                     appendCall.InvokedAsExtensionMethod,
                     appendCall.ArgsToParamsOpt,
                     appendCall.ResultKind,
                     appendCall.Type,
+                    temps,
                     appendCall);
 
-                Debug.Assert(usesBoolReturn == (rewrittenAppendCall.Type!.SpecialType == SpecialType.System_Boolean));
-                if (usesBoolReturn)
+                Debug.Assert(usesBoolReturn == (appendCall.Type!.SpecialType == SpecialType.System_Boolean));
+                resultExpressions.Add(rewrittenAppendCall);
+            }
+
+            if (usesBoolReturn)
+            {
+                // We assume non-bool returns if there was no parts to the string, and code below is predicated on that.
+                Debug.Assert(!node.Parts.IsEmpty);
+                // Start the sequence with appendProceedLocal, if appropriate
+                BoundExpression? currentExpression = appendShouldProceedLocal;
+
+                foreach (var appendCall in resultExpressions)
                 {
                     // previousAppendCalls && appendCall
-                    appendBinaryExpression = appendBinaryExpression is null
-                        ? rewrittenAppendCall
-                        : _factory.LogicalAnd(appendBinaryExpression, rewrittenAppendCall);
-                }
-                else
-                {
-                    builderPatternExpressions.Add(rewrittenAppendCall);
+                    currentExpression = currentExpression is null
+                        ? appendCall
+                        : _factory.LogicalAnd(currentExpression, appendCall);
                 }
+
+                resultExpressions.Clear();
+                resultExpressions.Add(handlerConstructionAssignment);
+
+                Debug.Assert(currentExpression != null);
+
+                var sequence = _factory.Sequence(
+                    appendShouldProceedLocal is not null
+                        ? ImmutableArray.Create(appendShouldProceedLocal.LocalSymbol)
+                        : ImmutableArray<LocalSymbol>.Empty,
+                    resultExpressions.ToImmutableAndClear(),
```

Comment:
nit: `ImmutableArray.Create(handlerConstructionAssignment)` since this is always exactly one expression?